mersenne-31/src/dft.rs

```diff
@@ -0,0 +1,256 @@
+///
+/// i.e. the even rows become the real parts and the odd rows become
+/// the imaginary parts.
+///
+/// This packing is suitable as input to a Fourier Transform over the
+/// domain Mersenne31Complex; it is inverse to `idft_postprocess()`
+/// below.
+fn dft_preprocess(
+    input: RowMajorMatrix<Mersenne31>,
+) -> RowMajorMatrix<Mersenne31Complex<Mersenne31>> {
+    assert!(input.height() % 2 == 0, "input height must be even");
+    RowMajorMatrix::new(
+        input
+            .rows()
+            .tuples()
+            .flat_map(|(row_0, row_1)| {
+                // For each pair of rows in input, convert each
+                // two-element column into a Mersenne31Complex
+                // treating the first row as the real part and the
+                // second row as the imaginary part.
+                row_0
+                    .iter()
+                    .zip(row_1)
+                    .map(|(&x, &y)| Mersenne31Complex::new(x, y))
+            })
+            .collect(),
+        input.width(),
+    )
+}
+
+/// Transform the result of applying the DFT to the packed
+/// `Mersenne31` values so that the convolution theorem holds.
+///
+/// Source: https://www.robinscheibler.org/2013/02/13/real-fft.html
+///
+/// NB: This function and `idft_preprocess()` are inverses.
+fn dft_postprocess(
+    input: RowMajorMatrix<Mersenne31Complex<Mersenne31>>,
+) -> RowMajorMatrix<Mersenne31Complex<Mersenne31>> {
+    let h = input.height();
+    let log2_h = log2_strict_usize(h); // checks that h is a power of two
+
+    // NB: The original real matrix had height 2h, hence log2(2h) = log2(h) + 1.
+    // omega is a 2h-th root of unity
+    let omega = Mersenne31Complex::primitive_root_of_unity(log2_h + 1);
+    let mut omega_j = omega;
+
+    let mut output = Vec::with_capacity((h + 1) * input.width());
+    output.extend(
+        input
+            .first_row()
+            .map(|x| Mersenne31Complex::new_real(x.real() + x.imag())),
+    );
+
+    for j in 1..h {
+        let row_x = input.row_slice(j);
+        let row_y = input.row_slice(h - j);
+
+        let row = row_x.iter().zip(row_y).map(|(&x, y)| {
+            let even = x + y.conjugate();
+            // odd = (x - y.conjugate()) * -i
+            let odd = Mersenne31Complex::new(x.imag() + y.imag(), y.real() - x.real());
+            (even + odd * omega_j).div_2exp_u64(1)
+        });
+        output.extend(row);
+        omega_j *= omega;
+    }
+
+    output.extend(
+        input
+            .first_row()
+            .map(|x| Mersenne31Complex::new_real(x.real() - x.imag())),
+    );
+    debug_assert_eq!(output.len(), (h + 1) * input.width());
+    RowMajorMatrix::new(output, input.width())
+}
+
+/// Undo the transform of the DFT matrix in `dft_postprocess()` so
+/// that the inverse DFT can be applied.
+///
+/// Source: https://www.robinscheibler.org/2013/02/13/real-fft.html
+///
+/// NB: This function and `dft_preprocess()` are inverses.
```

Comment:
```suggestion
/// NB: This function and `dft_postprocess()` are inverses.
```
If I understood correctly